### DISCUSSION TITLE
Prisma  Jest Testing for Dependency Injection and Functions without Context

### DISCUSSION THREAD

**Discussion by BenBausch** [prisma/prisma#23299] [Q&A]
### Question

I followed the guide here: https://www.prisma.io/docs/orm/prisma-client/testing/unit-testing#singleton. 

I have the following setup:

- Monorepo managed with NX
- Dependency Injection Pattern in my nest.js backend api
- Prisma ORM
- Testing using Jest

Current Testing Problem:
Trying to mock the prisma implementation inside a prisma interactive transaction.

My Question:
How to mock the prisma client for functions that do not have a context argument? As I can not provide the context to the function, the test does not use the mocked prisma implementations and tries to connect to the database.

Thank you very much in advance.

### Information about Prisma Schema, Client Queries and Environment (optional)

General Question Independent of the schema.

**Comment by ludralph** [comment]
Hi @BenBausch 👋 

Thank you for raising this question.

To mock the Prisma client for functions that do not have a context argument, you can use the singleton pattern as shown in the link you shared. This pattern allows you to mock the Prisma client without passing it as a parameter to your functions.
For example:
1. Create a `client.ts` file at your project's root and instantiate a Prisma Client instance:
```js
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()
export default prisma
```
2. Create a singleton.ts file at your project's root:
```js
import { PrismaClient } from '@prisma/client'
import { mockDeep, mockReset, DeepMockProxy } from 'jest-mock-extended'
import prisma from './client'

jest.mock('./client', () => ({
  __esModule: true,
  default: mockDeep<PrismaClient>(),
}))

beforeEach(() => {
  mockReset(prismaMock)
})

export const prismaMock = prisma as unknown as DeepMockProxy<PrismaClient>
```
This file tells Jest to mock a default export (the Prisma Client instance in ./client.ts), and uses the `mockDeep` method from `jest-mock-extended` to enable access to the objects and methods available on Prisma Client. It then resets the mocked instance before each test is run.

Are you running into any specific issue when trying to use the singleton pattern?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
